packages/grid/x-data-grid/src/internals/utils/useProps.ts

```diff
@@ -7,9 +7,43 @@ interface WithComponents {
   componentsProps?: GridSlotsComponentsProps;
 }
 
+/** Gathers props for the root element into a single `.forwardedProps` field */
+function groupForwardedProps<
+  T extends {
+    forwardedProps?: Record<string, any>;
+    [key: string]: any;
+  },
+>(props: T): T {
+  const keys = Object.keys(props);
+
+  if (
+    !keys.some((key) => key.startsWith('aria-')) &&
+    !keys.some((key) => key.startsWith('data-'))
+  ) {
+    return props;
+  }
+
+  const newProps = {} as Record<string, unknown>;
+  const forwardedProps: Record<string, unknown> = props.forwardedProps ?? {};
+
+  for (let i = 0; i < keys.length; i += 1) {
+    const key = keys[i];
+
+    if (key.startsWith('aria-') || key.startsWith('data-')) {
+      forwardedProps[key] = props[key];
+    } else {
+      newProps[key] = props[key];
+    }
+  }
+
+  newProps.forwardedProps = forwardedProps;
+
+  return newProps as T;
```

Comment:
Changing the order will possibly make it slightly more performant.
```suggestion
  if (
    keys.some((key) => key.startsWith('aria-')) ||
    keys.some((key) => key.startsWith('data-'))
  ) {
    const newProps = {} as Record<string, unknown>;
    const forwardedProps: Record<string, unknown> = props.forwardedProps ?? {};

    for (let i = 0; i < keys.length; i += 1) {
      const key = keys[i];

      if (key.startsWith('aria-') || key.startsWith('data-')) {
        forwardedProps[key] = props[key];
      } else {
        newProps[key] = props[key];
      }
    }

    newProps.forwardedProps = forwardedProps;
    return newProps as T;
  }

  return props;
```

Reply:
I didn't mean refer to it being performant due to frequency of one of the branches more than the other (if that is what you anticipated). The point was to avoid one `.some` in case the other evaluates to true by converting `&&` to `||`. For example, if `aria-*` attribute is there, we don't need to check for `data-*` as we handle it internally. (Or I am missing your point here 🤔 )

In any case, handling it in one `.some` would even be better.

P.S The shared article has some interesting insights 👍

Reply:
> I didn't refer to it being more performant due to the frequency of one of the branches more (or less) than the other (if that is what you anticipated). 

My bad, didn't picked up that change in the diff :> The change is not applicable anymore thanks to Andrew's suggestion though.